server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -230,6 +344,142 @@ private void abortMergeTask(MergeTask mergeTask) {
+        ) {
+            this.highStageWatermark = highStageWatermark;
+            this.highStageMaxHeadroom = highStageMaxHeadroom;
+            this.dataPaths = dataPaths;
+        }
+
+        @Override
+        public void run() {
+            FsInfo.Path leastAvailablePath = null;
+            IOException fsInfoException = null;
+            if (dataPaths == null) {
+                LOGGER.warn("Cannot read filesystem info because data path is not set in the env");
+                return;
+            }
+            for (NodeEnvironment.DataPath dataPath : dataPaths) {
+                try {
+                    FsInfo.Path fsInfo = getFSInfo(dataPath); // uncached
+                    if (leastAvailablePath == null || leastAvailablePath.getAvailable().getBytes() > fsInfo.getAvailable().getBytes()) {
+                        leastAvailablePath = fsInfo;
+                    }
+                } catch (IOException e) {
+                    if (fsInfoException == null) {
+                        fsInfoException = e;
+                    } else {
+                        fsInfoException.addSuppressed(e);
+                    }
+                }
+            }
+            if (fsInfoException != null) {
+                LOGGER.warn("unexpected exception reading filesystem info", fsInfoException);
+            }
+            if (leastAvailablePath == null) {
+                LOGGER.error("Cannot read filesystem info");
+                return;
+            }
+            // subtract disk space that already running merges are expected to fill
+            long leastAvailableDiskSpaceBytes = leastAvailablePath.getAvailable().getBytes();
+            for (MergeTask mergeTask : runningMergeTasks) {
+                leastAvailableDiskSpaceBytes -= mergeTask.estimatedRemainingMergeSize();
+            }
+            // also subtract the configured free disk space threshold
+            leastAvailableDiskSpaceBytes -= getFreeBytesThreshold(leastAvailablePath.getTotal(), highStageWatermark, highStageMaxHeadroom)
+                .getBytes();
+            // the rest is the maximum disk space available for a new merge task
+            long maxMergeSizeLimit = Math.max(0L, leastAvailableDiskSpaceBytes);
+            queuedMergeTasks.updateMaxPriorityLimit(maxMergeSizeLimit);
+        }
+
+        private static ByteSizeValue getFreeBytesThreshold(
+            ByteSizeValue total,
+            RelativeByteSizeValue watermark,
+            ByteSizeValue maxHeadroom
+        ) {
+            // If bytes are given, they can be readily returned as free bytes. If percentages are given, we need to calculate the free
+            // bytes.
+            if (watermark.isAbsolute()) {
+                return watermark.getAbsolute();
+            }
+            return ByteSizeValue.subtract(total, watermark.calculateValue(total, maxHeadroom));
+        }
+    }
+
+    static class PriorityBlockingQueueWithMaxLimit<E> {
```

Comment:
Perhaps we can add a comment on why we have this special queue. I think the primary benefit is that we can still process incoming merges that can be executed with this here, whereas if we did the blocking in `runMergeTask`, we'd not be able to do so once all threads are occupied.

Reply:
>whereas if we did the blocking in runMergeTask, we'd not be able to do so once all threads are occupied.
Yeah, this solution is not the best, keeping a thread stopped on a merge it selected some time ago (maybe the merge is not the smallest one anymore, and the same thread should instead be available to pick a different merge task).